noir-projects/aztec-nr/aztec/src/messages/message_delivery.nr

```diff
@@ -1,75 +1,165 @@
+    /// that the recipient cannot find it.
+    ///
+    /// An undecryptable or otherwise invalid note or event message will however simply be ignored by the recipient, who
+    /// can always validate the existence of the note or event on-chain.
+    ///
+    /// These guarantees make this delivery mechanism be quite similar to [MessageDeliveryEnum::OFFCHAIN], except the
+    /// sender does not need to establish an off-chain communication channel with the recipient, and neither party needs
+    /// to worry about backups.
+    ///
+    /// ## Costs
+    ///
+    /// Because the encrypted message is emitted on-chain as transaction private logs, this delivery option results in
+    /// transaction fees associated with DA gas. The length of the original message is irrelevant to this cost, since
+    /// all private logs are padded to the same length with random data to enhance privacy.
+    ///
+    /// However, no circuit gates are introduced when the message is encrypted. Therefore, on-chain unconstrained
+    /// messages do not affect proving time at all.
+    ///
+    /// ## Privacy
+    ///
+    /// No information is revelead on-chain about sender, recipient, or the message contents. The message itself reveals
+    /// no information about the sender or recipient, and requires knowledge of the recipient's private address keys in
+    /// order to obtain the plaintext.
+    ///
+    /// Delivering the message does produce on-chain information in the form of private logs, so transactions that
+    /// deliver many messages this way might be identifiable by their the large number of logs.
+    ///
+    /// Identifying that a log corresponds to a message between a given sender and recipient requires, among other
+    /// things, knowledge of both of their addresses **and** either the sender's or recipient's private address key.
+    pub ONCHAIN_UNCONSTRAINED: u8,
 
-    /// 3. Offchain
-    /// - Uses unconstrained encryption with offchain delivery
-    /// - Lowest cost since no onchain storage is needed and short proving times since no constraints are used
-    ///   for encryption
-    /// - Suitable when recipients can verify message validity through other means
-    /// - Requires setting up custom infrastructure for handling offchain delivery (e.g. cloud storage)
-    /// - Example: A payment app where a merchant receives the message offchain and is willing to release the goods
-    ///   once he verifies that the payment is correct (i.e. can decrypt the message and verify that it contains
-    ///   a legitimate token note - note with note commitment in the note hash tree).
-    pub UNCONSTRAINED_OFFCHAIN: u8,
+    /// Delivers the message on-chain, guaranteeing the recipient will receive the correct content.
+    ///
+    /// >**WARNING**: this delivery mode is currently NOT fully constrained. The log's tag is unconstrained, meaning a
+    /// malicious sender could manipulate it to prevent the recipient from finding the message.
```

Comment:
I would link here the issue - that will make it more straightforward to for people to go scream at us and less likely for us to remove this warning once this is implemented.

Not sure if [this one](https://github.com/AztecProtocol/aztec-packages/issues/14565) is still relevant.